File: NeuralNetworkPrimer/application.py
#######################################################################################################################
# A Primer on Neural Networks
#
# Paul Zivich (2023/11/30)
#######################################################################################################################

import numpy as np
import pandas as pd
from scipy.optimize import minimize
from delicatessen import MEstimator
from delicatessen.estimating_equations import ee_glm

from activations import identity, leaky_relu


#######################################################################################################################
# Reading in data

d = pd.read_csv("nhanes.csv")
d['intercept'] = 1

outcome = np.asarray(d['hdl'])
predictors = np.asarray(d[['age', 'female', 'weight', 'height']])

#######################################################################################################################
# Defining loss, objective functions


def mse(y, yhat):
    return np.sum((y - yhat)**2)


def objective(params, nn, loss_function, y):
    yhat = nn(params)
    return loss_function(y, yhat)


#######################################################################################################################
# One-layer neural network

def nn_one_layer(params):
    # Parameters for neural network
    intercept, coefficients = params[0], params[1:]

    # Output layer
    ws = np.dot(predictors, coefficients) + intercept   # Weighted sum of inputs and predictors
    yhat = identity(ws)                                 # Activation function (identity)

    # Returning the generated predictions from the neural network
    return yhat


# Training the neural network
starting_vals = np.array([0, 0, 0, 0, 0])             # Starting values for the optimization
opt = minimize(objective,                             # Objective (loss) function to minimize
               x0=starting_vals,                      # ... with provided starting values
               args=(nn_one_layer, mse, outcome),     # ... extra arguments passed to `objective`
               method='nelder-mead',                  # ... using the Nelder-Mead algorithm
               tol=1e-9,                              # ... setting maximum error tolerance till convergence
               options={"maxiter": 100000})           # ... increasing iterations to large number
# print(opt.message)                                  # Check that neural network found a minimum
est_params_nn1 = opt.x                                # Return estimated parameters for neural network
yhat_nn1 = nn_one_layer(est_params_nn1)               # Generating predictions for observations from neural network
print("Loss:", mse(outcome, yhat_nn1))
print("Neural Network Parameters")
print(est_params_nn1)

# Comparing neural network parameters to GLM
def psi(theta):
    return ee_glm(theta=theta, X=X, y=outcome,
                  distribution='normal',
                  link='identity')


X = d[['intercept', 'age', 'female', 'weight', 'height']]
estr = MEstimator(psi, init=starting_vals)            # Setup M-estimator
estr.estimate()                                       # Estimate parameters via root-finding
est_params_glm = estr.theta                           # Return estimated parameters by GLM
print("GLM Parameters")
print(est_params_glm)


#######################################################################################################################
# Two-layer neural network


def nn_two_layer(params):
    # Parameters for neural network
    intercept_h1_n1, coef_h1_n1 = params[0], params[1:5]
    intercept_h1_n2, coef_h1_n2 = params[5], params[6:10]
    intercept_out, coef_out = params[10], params[10:]

    # Hidden layer
    hn1 = np.dot(predictors, coef_h1_n1) + intercept_h1_n1  # Weighted sum of inputs and predictors for hidden node 1
    hn1 = leaky_relu(hn1, scale=0.01)                       # Activation function for hidden node 1
    hn2 = np.dot(predictors, coef_h1_n2) + intercept_h1_n2  # Weighted sum of inputs and predictors for hidden node 2
    hn2 = leaky_relu(hn2, scale=0.01)                       # Activation function for hidden node 2

    # Output layer
    ws = intercept_out + coef_out[0]*hn1 + coef_out[1]*hn2  # Weighted sum of inputs and predictors for output node
    output = identity(ws)                                   # Activation function for output node (identity)

    # Returning the generated predictions from the neural network
    return output


# Training the neural network
rng = np.random.RandomState(30032023)                 # Random state for pseudo-random number generator
starting_vals = rng.uniform(-5, 5, size=13)           # Starting values for optimization (random uniform over -5 to 5)
opt = minimize(objective,                             # Objective (loss) function to minimize
               x0=starting_vals,                      # ... with provided starting values
               args=(nn_two_layer, mse, outcome),     # ... extra arguments passed to `objective`
               method='nelder-mead',                  # ... using the Nelder-Mead algorithm
               tol=1e-9,                              # ... setting maximum error tolerance till convergence
               options={"maxiter": 100000})           # ... increasing iterations to large number

# print(opt.message)                                  # Check that neural network found a minimum
est_params_nn2 = opt.x                                # Return estimated parameters for neural network
yhat_nn2 = nn_two_layer(est_params_nn2)               # Generating predictions for observations from neural network

# Training the neural network, but with different starting values
rng = np.random.RandomState(30041993)                 # Random state for pseudo-random number generator
starting_vals = rng.uniform(-5, 5, size=13)           # Starting values for optimization (random uniform over -5 to 5)
opt = minimize(objective,                             # Objective (loss) function to minimize
               x0=starting_vals,                      # ... with provided starting values
               args=(nn_two_layer, mse, outcome),     # ... extra arguments passed to `objective`
               method='nelder-mead',                  # ... using the Nelder-Mead algorithm
               tol=1e-9,                              # ... setting maximum error tolerance till convergence
               options={"maxiter": 100000})           # ... increasing iterations to large number
# print(opt.message)                                  # Check that neural network found a minimum
print("Loss:", mse(outcome, yhat_nn2))
print("Loss:", mse(outcome, nn_two_layer(opt.x)))


#######################################################################################################################
# Three-layer neural network


def nn_three_layer(params):
    # Parameters for neural network
    intercept_h1_n1, coef_h1_n1 = params[0], params[1:5]
    intercept_h1_n2, coef_h1_n2 = params[5], params[6:8]
    intercept_h1_n3, coef_h1_n3 = params[8], params[9:11]
    intercept_h2_n1, coef_h2_n1 = params[11], params[12:14]
    intercept_h2_n2, coef_h2_n2 = params[14], params[15:17]
    intercept_out, coef_out = params[17], params[18:]

    # Hidden layer
    h1n1 = np.dot(predictors, coef_h1_n1) + intercept_h1_n1          # Weighted sum for hidden node 1
    h1n1 = leaky_relu(h1n1, scale=0.01)                              # Activation function for hidden node 1
    h1n2 = np.dot(predictors[:, 1:3], coef_h1_n2) + intercept_h1_n2  # Weighted sum for hidden node 2
    h1n2 = leaky_relu(h1n2, scale=0.01)                              # Activation function for hidden node 2
    h1n3 = np.dot(predictors[:, 2:], coef_h1_n3) + intercept_h1_n3   # Weighted sum for hidden node 3
    h1n3 = leaky_relu(h1n3, scale=0.01)                              # Activation function for hidden node 3

    # Hidden layer
    h2n1 = intercept_h2_n1 + coef_h2_n1[0]*h1n1 + coef_h2_n1[1]*h1n2  # Weighted sum for hidden node 1
    h2n1 = leaky_relu(h2n1, scale=0.01)                               # Activation function for hidden node 1
    h2n2 = intercept_h2_n2 + coef_h2_n2[0]*h1n2 + coef_h2_n2[1]*h1n3  # Weighted sum for hidden node 2
    h2n2 = leaky_relu(h2n2, scale=0.01)                               # Activation function for hidden node 2

    # Output layer
    ws = intercept_out + coef_out[0]*h2n1 + coef_out[1]*h2n2  # Weighted sum of inputs and predictors for output node
    output = identity(ws)                                     # Activation function for output node (identity)

    # Returning the generated predictions from the neural network
    return output


# Training the neural network
rng = np.random.RandomState(30032023)                 # Random state for pseudo-random number generator
starting_vals = rng.uniform(-5, 5, size=20)           # Starting values for optimization (random uniform over -5 to 5)
opt = minimize(objective,                             # Objective (loss) function to minimize
               x0=starting_vals,                      # ... with provided starting values
               args=(nn_three_layer, mse, outcome),   # ... extra arguments passed to `objective`
               method='nelder-mead',                  # ... using the Nelder-Mead algorithm
               tol=1e-9,                              # ... setting maximum error tolerance till convergence
               options={"maxiter": 100000})           # ... increasing iterations to large number
# print(opt.message)                                  # Check that neural network found a minimum
est_params_nn3 = opt.x                                # Return estimated parameters for neural network
yhat_nn3 = nn_three_layer(est_params_nn3)             # Generating predictions for observations from neural network
print("Loss:", mse(outcome, yhat_nn3))

File: NeuralNetworkPrimer/application.R
################################################################################
# A Primer on Neural Networks
# 
# Paul Zivich (2023/11/30)
################################################################################

library(nloptr)

################################################################################
# Reading in data

setwd('C:/Users/zivic/Documents/open-source/publications-code/NeuralNetworkPrimer')
d <- read.csv("nhanes.csv")

outcome = d$hdl
predictors = d[, c('age', 'female', 'weight', 'height')]


################################################################################
# Defining loss, activation, objective functions

mse <- function(y, yhat){
    # Mean squared error loss function
    return(sum((y - yhat)^2))
}

identity <- function(x){
    # Identity activation function
    return(x)
}

leaky_relu <- function(x, scale=0.01){
    # Leaky rectified linear unit
    xstar = ifelse(x < 0, scale*x, x)
    return(xstar)
}

objective <- function(params, nn, loss_function, y){
    # Generic objective function for neural networks
    yhat = nn(params)
    return(loss_function(y, yhat))
}

# Maximization algorithm arguments
nm_control <- list(maxeval=100000, xtol_rel=1e-9)

################################################################################
# One-layer neural network

nn_one_layer <- function(params){
    # Parameters for neural network
    intercept = params[1]
    coefficients = params[2:length(params)]

    # Output layer
    ws = as.matrix(predictors) %*% as.matrix(coefficients) + intercept
    yhat = identity(ws)

    # Returning the generated predictions from the neural network
    return(yhat)
}

# Training the neural network
starting_vals = c(0, 0, 0, 0, 0)                  # Starting values for optim
nn1_fit = nloptr::neldermead(starting_vals,       # Nelder-Mead algorithm
                             objective,           # ... with objective
                             nn=nn_one_layer,     # ... for neural net
                             loss_function=mse,   # ... with loss function
                             y=outcome,           # ... observed outcome
                             control=nm_control)  # ... specifications
nn1_fit$par                                       # Neural network parameters

# Loss function results
objective(nn1_fit$par, nn_two_layer, mse, outcome)

# Comparing to linear regression
flm = lm(hdl ~ age + female + weight + height, d)
flm$coefficients                                  # GLM parameters


################################################################################
# Two-layer neural network

nn_two_layer <- function(params){
    # Parameters for neural network
    intercept_h1_n1 = params[1]
    coeff_h1_n1 = params[2:5]
    intercept_h1_n2 = params[6]
    coeff_h1_n2 = params[7:10]
    intercept_out = params[11]
    coef_out = params[12:length(params)]
    
    # Hidden layer
    hn1 = as.matrix(predictors) %*% as.matrix(coeff_h1_n1) + intercept_h1_n1
    hn1 = leaky_relu(hn1, scale=0.01)
    hn2 = as.matrix(predictors) %*% as.matrix(coeff_h1_n2) + intercept_h1_n2
    hn2 = leaky_relu(hn2, scale=0.01)
    
    # Output layer
    ws = intercept_out + coef_out[1]*hn1 + coef_out[2]*hn2
    yhat = identity(ws)
    
    # Returning the generated predictions from the neural network
    return(yhat)
}

# Starting values from Python random draw
starting_vals = c(-4.85766562, -0.68064447, 4.78745109, 1.66116221, 2.39312546, 
                  3.37231983, 4.87319149, -3.08506736, -3.66220176, -3.75072766, 
                  -4.34286364, 2.32722006, 0.71739167)
nn2_fit = nloptr::neldermead(starting_vals,        # Nelder-Mead algorithm
                             objective,            # ... with objective
                             nn=nn_two_layer,      # ... for neural net
                             loss_function=mse,    # ... with loss function
                             y=outcome,            # ... observed outcome
                             control=nm_control)   #... specifications

# Loss function results
objective(nn2_fit$par, nn_two_layer, mse, outcome)  # 1049877

# Starting values from Python random draw
starting_vals = c(-0.12558097, 1.67103887, 2.79717236, -2.09890536, 
                  -2.81267949, -0.79644393, 3.23312734, 1.61378279, 
                  2.25609064, -3.16291761,  2.77545355, -2.84309898,
                  2.33620732)
nn2_fit = nloptr::neldermead(starting_vals,       # Nelder-Mead algorithm
                             objective,           # ... with objective
                             nn=nn_two_layer,     # ... for neural net
                             loss_function=mse,   # ... with loss function
                             y=outcome,           # ... observed outcome
                             control=nm_control)  # ... specifications

# Loss function results
objective(nn2_fit$par, nn_two_layer, mse, outcome)  # 1001958

################################################################################
# Three-layer neural network

nn_three_layer <- function(params){
    # Parameters for neural network
    intercept_h1_n1 = params[1]
    coeff_h1_n1 = params[2:5]
    intercept_h1_n2 = params[6]
    coeff_h1_n2 = params[7:8]
    intercept_h1_n3 = params[9]
    coeff_h1_n3 = params[10:11]
    intercept_h2_n1 = params[12]
    coeff_h2_n1 = params[13:14]
    intercept_h2_n2 = params[16]
    coeff_h2_n2 = params[16:17]
    intercetp_out = params[18]
    coef_out = params[19:length(params)]
    
    # Hidden layer
    h1n1 = as.matrix(predictors) %*% as.matrix(coeff_h1_n1) + intercept_h1_n1
    h1n1 = leaky_relu(h1n1, scale=0.01)
    h1n2 = as.matrix(predictors[,2:3]) %*% as.matrix(coeff_h1_n2) + intercept_h1_n2
    h1n2 = leaky_relu(h1n2, scale=0.01)
    h1n3 = as.matrix(predictors[,3:4]) %*% as.matrix(coeff_h1_n3) + intercept_h1_n3
    h1n3 = leaky_relu(h1n3, scale=0.01)
    
    # Hidden layer
    h2n1 = intercept_h2_n1 + coeff_h2_n1[1]*h1n1 + coeff_h2_n1[2]*h1n2
    h2n1 = leaky_relu(h2n1, scale=0.01)
    h2n2 = intercept_h2_n2 + coeff_h2_n2[1]*h1n2 + coeff_h2_n2[2]*h1n3
    h2n2 = leaky_relu(h2n2, scale=0.01)
    
    # Output layer
    ws = intercetp_out + coef_out[1]*h2n1 + coef_out[2]*h2n2
    yhat = identity(ws)
    
    # Returning the generated predictions from the neural network
    return(yhat)
}

# Starting values from Python random draw
starting_vals = c(-4.85766562, -0.68064447, 4.78745109, 1.66116221, 2.39312546,
                  3.37231983, 4.87319149, -3.08506736, -3.66220176, -3.75072766, 
                  -4.34286364, 2.32722006, 0.71739167, -2.07179524, -4.87551265, 
                  -0.76882816, 3.13232629, 3.46530322, 1.1681697, -2.12807223)
nn3_fit = nloptr::neldermead(starting_vals,       # Nelder-Mead algorithm
                             objective,           # ... with objective
                             nn=nn_three_layer,   # ... for neural net
                             loss_function=mse,   # ... with loss function
                             y=outcome,           # ... observed outcome
                             control=nm_control)  # ... specifications

# Loss function results
objective(nn3_fit$par, nn_three_layer, mse, outcome)
